### DISCUSSION TITLE
3436/7 Controllers for 6400

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#130] [General]
Reading the 6400 manual..

Maybe emulating the drums in RAM with OPTIONAL file backing, and MANDATORY persistence save?  SSD friendly.

Deviation from orig is that unless file backing is selected, crash recovery would not be possible.  Graceful shutdown would save them.  Oh, and it would be insanely fast.  Programs relying on access times would require modification.  Unless this is a larger issue than I think, I see no need for clock accurate emulation.

Even a somewhat tight pi zero should be able to handle at least one, if not more drums in RAM.

The 3437 could use a shared memory area, or even a socket pair (local and remote variants).  The 3436 operating from RAM entirely, the 3437 being an extension of it.

Anyone like the idea?  It would expand the usability of the 6k emulation .

**Comment by wschaub** [comment]
What software is currently able to use drums? 
I think KRONOS might have a driver for it but I don't know about NOS1, pretty sure NOS2 doesn't

it might be interesting to play with if I had an operating system that can use it. I'm already tinkering with 3000 series 7 track tape support.

**Comment by KC5VDJ** [comment]
Not a freaking clue, but it would be fun to play with.

Intersystem comms by way of dual-porting drums looks interesting for the hobbyist.

I am thinking by way of what I was just musing on that any dual-porting was by way of one-offs done in-house / under contract...potentially highly custom software / system config...

**Comment by kej715** [comment]
To create a DtCyber driver, we would need a hardware manual specifying
controller functions and/or PP source code enabling reverse engineering of
the controller functions.

**Comment by KC5VDJ** [comment]
60333100A 😁

Doesn't look that hard to deal with...

Three four bit SIPO shifters to assemble the byte on the 865 clocked on the fourth bit for ready?  Efficient in hardware.

**Comment by KC5VDJ** [comment]
Do any of these drums still exist? Out of those how many are there that might have data anyone actually cares about?  I find it possible, but highly unlikely that anyone would ever do an actual format accurate reproduction of one, especially with the traditional transient nature of drum data, so I'll just present linear RAM (optionally file backed but not SSD safe), and make no attempt at format accuracy, just command / response / status accuracy.

Angles can be derived? Might not have read far enough...

**Comment by kej715** [comment]
It's a 3000 series device, so it requires a data converter such as a 6681
for interaction with the 6400 (or 70/170/800). That will make the DtCyber
driver a bit more complicated, and similar to interacting with the other
3000 series equipment already supported.

**Comment by KC5VDJ** [comment]
Kinda gathered that looking at things, but overall, I think the driver would be a relatively simple one for the type.  A good one to start with maybe.  Being new support, nobody gets hurt..  of course, imported drum images would have to be byte accurate, not format accurate, but again, other than existing software which would work fine if saved to tape or something, but again, bitrot....  Any user getting hurt would be obviously anyone relying on drums already in their program but without a byte accurate image, and an extreme edge case where a drum had to be reconstructed...best they write their own tools for that, data recovery would require more than just parity.

**Comment by KC5VDJ** [comment]
Yeah, the note on p30 is what I was thinking.  Sort of a scatter / gather thing.  I will derive the angles if no table.  It also implies a queue / request table of some kind, such as an os would use for the purpose.  Yes, 0044 needs to work.

**Comment by wschaub** [comment]
I did a quick scan through my collection of source code and it looks like the only operating system that has a driver for the drum is KRONOS2.1  I tried to find it in NOS/BE or HUSTLER but came up empty handed, if it is in there it's non-obvious where the driver code might live.  I thought it should be 1SP but didn't find anything in there. 

Did you want to take a stab at doing a driver? I'm currently tied up with mt362x.c making it actually usable with KRONOS2.1 and hopefully NOS1.3/NOS1.4 as well.  (with a mind towards using it for interchange with other emulators that recognize 7track tape formats and possibly reading some interesting tapes I found in .BCD format at https://www.piercefuller.com/library/cdc.html ) 

you can easily test out your driver code using the LDC,nnnn. FNCnn,nnnn. DCNnn.  ACNnn.  IANnn. and OANnn. console commands in KRONOS or NOS to interact with your device without writing PP code.

If you want to write test PP code I have a sort of skeleton of a CPU program and a PP helper program that gets called by the CPU program that you can insert code that interacts with a driver into if you want it. 

COS might have some support in itas well  but it's hard to reason about COS for me.  KRONOS2.1 is functional enough that it should be a fairly easy time working with its mass storage driver for drums which is in the source deck 6DC. The KRONOS2.1 workshop and to a large degree the NOS1.4 IMS should get you up to speed on how the PP resident, SETMS macros and the storage driver overlays (of which 6DC is one) interact.  

On KRONOS2.1 substitute SYSTEXT for when it talks about NOSTEXT. I have jobs for building SYSTEXT and a few other reference decks like PPR, etc to make a listing.

**Comment by wschaub** [comment]
[6dc-listing.txt.gz](https://github.com/kej715/DtCyber/files/14552359/6dc-listing.txt.gz)

**Comment by KC5VDJ** [comment]
I might just...  Started with COS and how anemic it looked when compared with device support and examples in the (SIPROS-oriented) 6400 manual...  Adding it for others could be beneficial, and helpful (debug).

And yeah..drums...but, implemented in modern RAM...they were small...  At least for hobbyist / new effort on anything but Kronos and maybe cos...the construct seems really usable.

On a side, but related note, I read a document last night complaining heavily about the whole SIPROS / COS thing, and their HEAVY investment in SIPROS... CERN Computer Newsletter #2.  Interesting reading.

Oh, and grep finished in nos2...  One reference in cml3/uplrc/history to the 863 / 865 drums.  Then 1tn in nos 2.8.7 references the answer back drum in a terminal.  Then multiple references to J. J. Drummond who changed several files in 77/78.  Nos2 support would be new.

**Comment by kej715** [comment]
Based upon William's investigation so far, it seems that only KRONOS 1.2
supports the drum device. Apart from discovering that a PP mass storage
driver named 6DC in KRONOS supports the device, we also need to determine
how the device is referenced in the KRONOS CMR deck so that a system can
actually be configured to use it. As KRONOS apparently supported it, it
would be interesting to add support in DtCyber from a
historical perspective. However, I'll bet that very few CDC customers
actually used drums with 6000 series machines. And, if any did use them,
they probably used them only as rollout devices. Their capacity was limited
(about 400K 60-bit words), and their transfer rates weren't very impressive
(5 or 10 usec per 60-bit word, depending upon drum type). In contrast,
ECS/ESM had greater capacity and much greater transfer rates. ECS/ESM was
vastly superior to a drum as a rollout device from a
performance perspective, and it was probably superior from a
performance/cost perspective too.

Attempting to support drums in other operating systems, e.g., NOS 1 or NOS
2, would require more than just having a PP mass storage driver for it.
PP's such as SET would also need to support the device type. Specifically,
SET would need to recognize a 2-character device code for it, and it would
need to understand that it's a mass storage device capable of being used
for rollout, temp files, etc.

The University of Massachusetts (UMass) operated a CDC 3600 and CDC 3800
until 1974. The 3800 ran a homemade, timeshared operating system supported
by a 3000 series drum. The drum was used as a swapping device. The pair of
3000 series mainframes were replaced by a Cyber 74 in 1974, and the Cyber
74 ran KRONOS at first. The drum was not retained for use on the Cyber 74.
Instead, 844 disks were initially used for both permanent and temporary
file storage. The aggregate throughput of multiple 844's outperformed a
drum, and a single 844's capacity was much greater than a drum.

cheers,
Kevin

**Comment by pa0gri** [comment]
Kevin,

 

Most drums were used on 300 series systems.

When migrating to 6000 they kept their drum for a while.

 

Scope 3.3 also could use the drum. We had one site here. (secret…)

 

Gerard.

**Comment by KC5VDJ** [comment]
Well as far as I can tell, in the CML3 tools, with MALET, in the HISTORY file, PSR CMLA028 references removing the 501, 808, 813/814, 821, 853/854, ***863/865***, ***6603***, and LCC/M1000 on 1983/01/19.

The two letters apparently WERE "DC", which was reassigned when the drums were obsoleted, according to the comment.

**Comment by KC5VDJ** [comment]
Inter-system communications before networking...

On a PC, OS overhead would be lower not going through the hosted OS protocol stack...  Implemented in RAM...Doing an 844 in RAM would be impractical on most end-user systems and controllers/SBCs, simply due to size.

I think I will just for shits and giggles.  The 6400 System Bulletin (Dec. 1964) interestingly only mentions the 3436 and 861.  Still...  The 3487 and 863 support would be transparent to the hosted OS supporting the former.

![60333100A_3436_3437_Drum_Storage_Ctlr_Oct70-Page 08-Figure 1_082808](https://github.com/kej715/DtCyber/assets/136194499/cc2117ac-1f18-4ac4-8497-5f714f65df22)

Original max latency was 35mS, average 17.  Better or equivalent to modern networking.  Like I said, for hobbyist and student use....some neat stuff can be done.  Like having a dual-port 2Mx12 or 4Mx12 RAM to interface machines, with RAM latency (when not file backed, other than persistence at init / shutdown).

Yeah, ECS did it later, but...

**Comment by wschaub** [comment]
I think the 6681 is a pretty cool device overall from what I've learned about it so far. being able to easily integrate 3000 series devices must have been an overall win for CDC I imagine, at least for sites that still had 3000 series controllers for things like tapes and disks.  

A lot of 3000 series devices look very similar on a  6681 in terms of how you use them. there's generalized function codes for read and write operations, as well as a way to pass connect and device specific codes to the 3000 series devices, a way to handle interrupt status from the devices, etc. 

ECS was probably the most widely used for inter-mainframe communications at a single site. You have block transfers built right into the CPU itself in addition to DDP equipment that the PPs can use to talk to ECS.  

There's also the satellite coupler that is essentially what IBM would call a channel to channel adapter (that one isn't in dtcyber and as far as I know was only used by the SCOPE2 station software)

**Comment by KC5VDJ** [comment]
The whole ECS thing after 6600 serial number 7 might have been the reason drums were more legacy even then.  When exactly did the ECS multi mainframe interface appear?  With ECS, or was there a delay before someone did?

**Comment by kej715** [comment]
Yes, the 6681 was a very important device. In addition to supporting tapes
and disks, it was also used in connecting line printers, card readers, card
punches, and telecommunications equipment such as the DSA 311 (aka 3266
mux). DtCyber supports all of these.

DtCyber doesn't currently support the satellite coupler, mainly because
there's currently nothing to connect on the other side. Perhaps if DtCyber
supported the 7600, and we had a copy of SCOPE2 to run on it, we could
implement support for the satellite coupler and get the SCOPE Station
software working. DtCyber currently supports the Cray Station interface
which was probably influenced strongly by the SCOPE Station interface (in
fact, I wouldn't be surprised if many of the same people were involved in
implementing both originally).

Regarding ECS/ESM as a communications device ... while it certainly could
be used for generalized inter-mainframe communication, I'm not aware of any
standard CDC-developed solution that used it that way. MMF (Multi-Mainframe
Feature) used it for very specialized inter-mainframe communication; mainly
to coordinate access to shared disks and to detect which machines were up
and running. We operated an MMF configuration at UMass. A Cyber 175 and 730
shared a subset of the disk farm, and ESM enabled that to be optimized. It
was also possible to coordinate disk sharing through shared controller
interlocks, but interlocking via ESM was much faster.  Disks holding
permanent files were shared between the machines, and some temp file
devices were shared as well. For example, we used the queued file utilities
to dequeue output files from the 730 and stage them to shared disks where
the 175 could pull them down to its output queue. The line printers were
physically connected to the 175, so this process enabled output produced on
the 730 to become visible to the 175 which could actually print it.

cheers,
Kevin

**Comment by kej715** [comment]
The formal, CDC-maintained "MMF" feature probably didn't appear until NOS
1. As mentioned earlier, "MMF" was not a generalized inter-mainframe
communication solution (i.e., it was not a generalized networking
solution). Remember that ECS/ESM is limited to being shared by a maximum of
four mainframes, so though it could have been used to support generalized
networking, the maximum network size would have been rather small. ;-)

**Comment by KC5VDJ** [comment]
As for the satellite coupler (6682), that would be something for later, but the hw ref, diagrams, and cordwood dox are all in the collection.  Was noting the CE / diagrams manual for the 6681 also contains several goodies.  Heh.  Absurd idea, but enough for an asic or fpga?

**Comment by KC5VDJ** [comment]
Was thinking about any actual 3000 <-> 6000 possible situations irl maybe...

IIRC, CERN ran both simultaneously. No idea on interconnection if any, but if they did, barring bitrot, there may be code there.  Listings might be more intact if any exist (fiche?) I'm mainly talking about pp.

They MIGHT also have SCOPE 2 laying around.  Doubt Barking Sands would, even if they did have it laying around (how many 7700s were even sold?)

1. On an absurd note...  On a 16-64 CPU system...  How many 6400s and how many 3437Bs and 861/865/865s needed to recreate the opening of "Colossus.."? Oh, they used so many!

**Comment by nickglazzard1959** [comment]
CERN did connect their 3000 series and 6000 series machines. In fact, the
3000 machines were a key part of their early networking efforts starting in
the later 1960s from what I've read. It seems that CERN wrote their own
operating system for the 3000 machines, called FOCUS with "60,000 lines" of
code. This all comes from the document attached (1974) and a later
document written by P. Zanella in 1990. One of the diagrams in the 1974
document mentions a drum ... :-) ... but I'm not sure if it was attached to
the 3000 or 6000 system. It would be very interesting if FOCUS still
existed, but I would be very surprised if it did.

**Comment by nickglazzard1959** [comment]
Some more FOCUS information here:
https://cds.cern.ch/record/350964/files/551.pdf, here:
https://core.ac.uk/download/pdf/44077693.pdf and here:
https://core.ac.uk/download/pdf/44077693.pdf

**Reply by KC5VDJ** [reply]
Interesting stuff there.  The third link a dupe of the 2nd though.  No mention of a drum, but those were decent distances for the day.  

Belden always was trash, but cheap for short runs.  Back in the day, I ran a parallel network out of some dumpster dive coils, was running about T1, maybe a touch less, between the printer ports on a Pentium and dual Pentium 2.  That was a 100 foot run.  Granted I did the connectors myself, and soldered the crimps, and the ground to shield wire. FBSD 2 & 3 days.  I can't imagine that speed with any form of decent BER over the runs mentioned in the papers over the cables I used.  I would get occasional errors, not many.  Never did a proper BERT on it.  Had the tools at work, but Sprint kinda frowned on such for personal unless work-related, so error indication was by way of netstat, etc...  Still my soldering is quite good, my mom was Collins trained, indeed so good, Mr. Collins sent her to Fresno where she was Apollo trained.  Every F-16 from before she retired, basically all of them, almost, have her work in there. GD->Elbit.  Not just Falcons.  I had a good teacher.

Interestingly, my grounding was good enough to not be influenced by a couple of three foot diameter HF transmitting loops (5 to about 120 watts cw/rtty/pep) and a two meter halfwave (20-150 watts) less than 30 feet away on the porch, E and H intersecting the cable, at least on the near end, if not the whole thing, depending on band.  I didn't notice any issues from HF/VHF induced CURRENT from near-field H, but Mr. Maxwell says it was there, and I'll trust him on that.  The beacon packets through R0MIR "FreeBSD r0x da world!" When in range of the KC metro, were mine...or was it w3rld? I was into l4m3sp34k then.  I did have ferrites for common mode.

Terabit Ethernet still work in progress right now? 😁😎. (Last I checked)

**Comment by KC5VDJ** [comment]
Hmmm. Gordon Bell mentions the drums being used(?) with the 7600.  I think he got his capacity number wrong, unless they made new drum models of the type in '69.  I'm guessing he refers to the 863?  Again, SCOPE 2.

https://gordonbell.azurewebsites.net/craytalk/sld055.htm

**Comment by kej715** [comment]
It might have been relatively common for sites to implement their own
operating systems for 3000 series mainframes. I mentioned previously that
UMass operated a 3600 and 3800 prior to replacing them with a Cyber 74. The
3600 ran CDC's SCOPE operating system (yes, the CDC-developed OS for 3000's
was named SCOPE too) and was used as a batch-only machine. The 3800 ran an
interactive operating system developed by the university, and cutely named
UMAS (Unlimited Machine Access for Students). In addition to implementing
an OS for the 3800, the university also implemented an APL interpreter
which was widely distributed to other academic sites running 3000's. Later,
the same small team that implemented APL for the 3800 implemented it for
the Cyber 74. This became CDC's APL 2. I joined the team in '78, and we
developed APL/VE under contract with CDC for the Cyber 180 series.

cheers,
Kevin

**Reply by KC5VDJ** [reply]
"unlimited" was a bit pretentious 😁

One of these days, I'd love to lay my hands on a pair of Selectrics, correspondence and APL.  Gotta admire those machines, works of art.

In a way, so was APL.  I'd like to learn it sometime.

**Comment by KC5VDJ** [comment]
Oh freaking wow!  What does the 8600 look like?

https://gordonbell.azurewebsites.net/craytalk/sld057.htm

It took how much longer to put something worthy into the form?  Norris still liked the idea ($).

Heh! 13 years! Didn't Point Mugu eventually get the appropriation for it?  Still pure Cyber when I worked there, but the cray was in every year's budget request, after the NASA report on the ETA (I still wonder if the memory design on the latter was influenced by plan 9, mach, domainOS, etc.. The way too early optical interface implies the possibility, all of which in the optical domain could potentially be used in optical lattice memory, thus the os concepts.  The switching of such is large scale and clunky today even.  Yeah yeah yeah, the LOC in a sugar cube of bacteriorhodopsin crystals...how do you address and use it...gotta love the 80s.  Clarke and others who knew Licklider were somehow convinced that very small optical switching on a large scale would be possible in some reasonable amount of time.  Fiegenbaum's paranoia too...

**Reply by KC5VDJ** [reply]
![img057](https://github.com/kej715/DtCyber/assets/136194499/3b2b7363-5bc7-4f8b-84ec-a034ca614469)

**Comment by kej715** [comment]
The preliminary reference manual for the 8600 can be found here on
Bitsavers ...

     http://www.bitsavers.org/pdf/cdc/8600/Preliminary8600RefMan_Aug72.pdf

**Reply by KC5VDJ** [reply]
Heh, he fixed negative zero!

**Comment by KC5VDJ** [comment]
Makes one wonder if his cooling guru got fluerinert going for the 8600...  I would imagine a lot of work being done at the time on self-contained cooling systems on rolling platforms in thermally challenging environments....patents?

**Comment by nickglazzard1959** [comment]
Oops ... apologies for duplicate link and documents not being attached (originally done via email rather than directly in Github, so I'll blame that). Here is a Google Drive link to an archive of PDFs that mention CDC 3100/3200 systems used in CERN's ancient FOCUS network (retired in 1978, probably). https://drive.google.com/file/d/15KrPAmssKHbDhRxb0YL7WJyeaG6gwtOb/view?usp=sharing
The T. Bloch one does mention a drum, but only in passing, and probably just attached to the 6600.
Yes, it sounds like a lot of people rolled their own operating systems for CDC 3000 series machines. UMAS sounds interesting. They also certainly had very pretty consoles from photos I've seen. It doesn't look like much (if any) software has survived for them, though. I'd love to be wrong about that. 
APL is really important historically ... so many concepts and terminology came from that and can still be seen in array processing systems like NumPy today. Still fun to use APL on NOS  ... :-)

**Comment by nickglazzard1959** [comment]
Some 8600 related links you may have seen before: https://www.facebook.com/TheSupercomputers/photos/pb.100039709280314.-2207520000/1355286474981126/?type=3
https://www.facebook.com/TheSupercomputers/photos/pb.100039709280314.-2207520000/1169265643583211/?type=3
https://www.facebook.com/TheSupercomputers/photos/pb.100039709280314.-2207520000/1167877253722050/?type=3
Also some nice renderings of 3000 series machines there.

**Reply by KC5VDJ** [reply]
Interesting stuff!  Gotta love his design ethos.

Not on social media, got tired of the  terrorism.

**Comment by KC5VDJ** [comment]
Unless someone has a better suggestion, I propose the device type of "ZD" in NOS1/2 based on the comment in the NOS 2 CE tools about "DC" being reassigned in 1983.

In the admittedly unlikely case of any legacy site sources for NOS1/2 that utilize the drum device as "DC" that still exist and have an actual need to be run again, they will need to be changed to refer to "ZD" (D for Drum).  The fact is that the CE tools notation was in 1983.  It had to be supported for a while, or it would have been gone sooner.  Someone was still using them, at least until the early 80s.  Lemme guess, 20 year contract under FAR for a highly custom config?

Unknown status in BE.  I would assume the change was across the board.

Were Scope and Kronos still being maintained in 1983 (officially)?  If not, changes there would be unnecessary.

**Reply by KC5VDJ** [reply]
I might be open to "ZC" (C for cylinder, and for single character source change (effecting NOS only)).  The argument against being a mass storage nomenclature thing.  Cylinder means something else.

**Reply by KC5VDJ** [reply]
After this, I might do the 3234 just for the novelty.  Hmmm... Each strip in its own file in a hierarchial directory tree? 😁😎

**Comment by KC5VDJ** [comment]
Question: before I dig down, in DevSlot, in types.h, is PCI referring to PCI?  I know DEC did (at least to chassis of 256 PCI cards).

**Comment by kej715** [comment]
Yes, PCI refers to actual PCI. However, it does -not- imply that CDC
machines supported PCI. Instead, some members of the DtCyber community have
PCI cards in their PC's that provide interfaces to CDC peripherals. These
PCI cards present Cyber channel interfaces. They enable DtCyber to drive
actual CDC peripherals, mainly CDCNet DI's, although it's possible that
some members might have other types of old CDC peripheral devices.

**Comment by kej715** [comment]
As previously discussed, the drum was a 3000 series peripheral and probably
was dual-ported. Most 6000 series peripherals, especially disks, were
dual-ported, as were most 6000 series controllers. This feature was
leveraged in two ways:

   1. It could be used to provide a single mainframe with more than one
   path to each disk, thus maximizing aggregate disk I/O throughput.
   2. It could be used to enable two mainframes to share disks.

Possibly, a dual-ported drum could be used in the same two ways, although
it might not be known whether the dual-ported feature of a drum was ever
actually used on a 6000 series mainframe.

**Comment by KC5VDJ** [comment]
That makes sense with the PCI thing.  I've seen old vectorboard brand ISA wire-wrap boards done up as channel(?) interfaces for Zenith 248s (odd polled-only serial port, excellant Alps keyswitches).

As for if dual-ported drums were used, I have no clue, but the diagram seems to imply intent, as does the command structure.  I'll do it as illustrated in the manual.  3436-A representing a single side, single connection (compat with Kronos 6dc); the 3637-A supporting dual channels; the 3637-B being dual channel with 865 drum capability (sectors instead of bytes, and two status bits taking on different meaning).

I'm using the card punch driver as a template (cp3446.c).  It seems to have most of what I need after stripping it, it also keeps me from introducing a clashing coding style (I prefer indented scoping and switches, etc, a little differently).

I'm kinda looking at this from the hobbyist / CS student / etc perspective, not necessarily the legacy code aspect.  Heck, I'm even thinking about doing the strip storage just out of curiosity and novelty.  The 3681-A, by way of a 6881, as another example, would give access to at least an interface for devices all the way back to the 160 and 1604, with one "new" CPU "addition" to DtCyber.

I have three simple rules:  Don't break the mainframe; the interface must be as specified to not break existing code, should any be found; and if operating hardware were found, it also must work with any PP drivers (museums sometimes turn things on).

**Reply by KC5VDJ** [reply]
Of course that is the channel visible interface.

**Reply by wschaub** [reply]
Another option for playing around with unsupported devices would be to do something like this:
set up an EST entry set to OFF or DOWN this will allow you to reserve the channel. then you can write a regular PP program to talk to the device. and a CPU program that will call it for you. the CPU and PP program can communicate via memory in an address passed to the PP program.  

In practice even the EST entry doesn't matter much as long as you know that channel isn't going to be used by the operating system.  it amounts to writing a driver in practice anyway but your new helper PP doesn't need to integrate with the operating system from day one this way.  I'm attaching a CPU and PP program I used to play with the mass storage drivers.  it started out as an idea to create a dtcyber device that could be used to control dtcyber from inside dtcyber via control card calls but never got to be more than a stub. the "dummy" routine I've just been filling in to do whatever for me and then I write back anything I need to capture to the CPU program which I then have dump memory to view it. 

just sysedit the resulting binaries into the system and the "dtcyber" command will call the helper PP fill it in with code to play with your drum and you should have the start of something that can exercise your device a bit. 

configuring the drum on something like KRONOS2.1 would be very helpful in developing the driver overall though since you can log all the operating system's interactions with the device and feel out how it's really supposed to act. there's a lot of non-obvious things different operating systems do with devices that doesn't make sense with what the device manuals say sometimes. 

[PPtest_skeleton_job.txt](https://github.com/kej715/DtCyber/files/14704463/PPtest_skeleton_job.txt)

**Reply by KC5VDJ** [reply]
Yeah, I'll definitely have to get Kronos going again for testing.  Thanks for the code as well, it'll come in handy.

**Comment by kej715** [comment]
Regarding coding style, please read the CONTRIBUTING.md
<https://github.com/kej715/DtCyber/blob/main/CONTRIBUTING.md> file, if you
haven't read it already. Tools are available to help comply with the coding
style requirements. The coding style was established by the original author
of DtCyber, Tom Hunter.

**Reply by KC5VDJ** [reply]
Thanks.

**Comment by KC5VDJ** [comment]
Thinking about the pp drivers...  In the unlikely event of more than one working one...  I guess the stagger start would definitely be a thing, maybe a power conditioner as well.  I would have to have a flag for bypassing stagger start.

By the way, weren't drums manufactured at the old scotch tape factory on 7th & Minnehaha, or next to it?  Right by a decent taqueria, last I was in Saint Paul.

**Reply by KC5VDJ** [reply]
Actually, using modern LF-MF power conditioning, maybe Li-ion or Li-Po backed, they would be small, only subject to wire size and distance, but in theory, stagger-start might not have to be two hours, but more like 22 or so minutes (one minute for the os to get its thing going on each in sequence, after fifteen minutes spinup), and individual power conditioners / isolation for each might fit in a cabinet (?) or a small auxillary cabinet.  Of course, that option would require stable power and wire rated for surge.  Modern switching supplies cure a lot of ills.  Like radio restoration, modern discretes are acceptable "under the hood".  Same idea.  It isn't too hard to get ripple to die for, just takes cash.  If there is one thing I wouldn't trust in a restoration would be the power supplies and filtration, and if they used tantalum caps for bypass, ugh... I'd recommend the Hakko FR-301, with the proper tips.  Like a champ, but keep it clean ($20-30+ tips).  Then again, if they did use tantalum, and most did, most of those are currently unsafe, meaning they all have to come out (and if not, I'd not turn it on without a halon system at the ready) maybe one of the Hakko stations would be better because of volume.  Then again the FR-301 is quite capable, and portable.  That would be fun, recapping the bypasses on a 170...how many?  With radios, I only trust silver-mica, silver-ceramic (depending on wear), and air / vacuum, and other reliable types, every other cap has to go...fu*k fires.  I'd imagine recapping would be a major expense.

**Comment by KC5VDJ** [comment]
The bit about the 865 is throwing me a bit...  If no point, why implement it in germanium?

If it will interrupt at start of sector, leaving no time for operation on that sector.... I have to be missing something here.  

Germanium in TO-5(?) cans and cordwood.  Why was this implemented on the 865?

Did this get used in a "fudged" pre-read fashion?  Set it a sector before then count cycles and you are at the next?  Seems convoluted when that coulda been also in germanium.

![Acrobat_3evvjp1ZHF_090932](https://github.com/kej715/DtCyber/assets/136194499/c02cc4d0-439d-4e17-9ccf-59ca188c9c2d)
